### ISSUE TITLE
Tracking issue for optional `io::Read`/`io::Write` methods

### ISSUE THREAD

**Issue by thaliaarchi** [rust-lang/rust#136756]
This tracks the status of supplying optional methods in implementations of `io::Read` and `io::Write` in the standard library.

`Read` and `Write` have many methods which are provided by default, and many implementors do not override them, which can be problematic in some cases. For example, `StdoutLock<'_>` does not implement `write_fmt`, so it reacquires the lock on every inner write. Platform abstractions for stdio are particularly lacking; most of which do not have `read_buf`, so the buffer is forced to be initialized, and some platforms which support it might be missing vectored read/write. Since the standard library is general, this should be consistent.

I am working on addressing these gaps and have created this issue to track progress for myself and reviewers. I am still a new contributor to Rust, so please let me know if I have missed something about the process.

This surveys of all implementations of `io::Read` and `io::Write` in `rust-lang/rust`, including for private types, by searching for for `\b(Read|Write) for ` in .rs files. This should have no false negatives, as opposed to viewing references, which would omit other targets.

## `Read`

```rust
pub trait Read {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
    fn read_buf(&mut self, buf: BorrowedCursor<'_>) -> Result<()> { … }
    fn read_buf_exact(&mut self, buf: BorrowedCursor<'_>) -> Result<()> { … }
    fn read_vectored(&mut self, bufs: &mut [IoSliceMut<'_>]) -> Result<usize> { … }
    fn is_read_vectored(&self) -> bool { … }
    fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> { … }
    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<usize> { … }
    fn read_to_string(&mut self, buf: &mut String) -> Result<usize> { … }
}
```

Columns:
- `buf`: `read_buf`
- `buf_exact`: `read_buf_exact`
- `vectored`: `read_vectored` and `is_read_vectored`
- `exact`: `read_exact`
- `to_end`: `read_to_end`
- `to_string`: `read_to_string`

| `impl Read for T`                                        | `buf` | `buf_exact` | `vectored` | `exact` | `to_end` | `to_string` |
| -------------------------------------------------------- | ----- | ----------- | ---------- | ------- | -------- | ----------- |
| [`&mut Read`]                                            | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`Box<Read>`]                                            | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`&[u8]`][&slice]                                        | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`VecDeque<u8>`][VecDeque as Read]                       | ✅     | ✅           |            | ✅       | ✅        | ✅           |
| [`std::io::Chain<Read, Read>`]                           | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`std::io::Take<Read>`]                                  | ✅     |             |            |         |          |             |
| [`std::io::Empty`][Empty as Read]                        | ✅     |             |            |         |          |             |
| [`std::io::Repeat`]                                      | ✅     |             | ✅          |         | ✅        | ✅           |
| [`std::fs::File`][File as Read]                          | ✅     |             | ✅          |         | ✅        | ✅           |
| [`&std::fs::File`][&File as Read]                        | ✅     |             | ✅          |         | ✅        | ✅           |
| [`Arc<std::fs::File>`][Arc<File> as Read]                | ✅     |             | ✅          |         | ✅        | ✅           |
| [`std::io::Cursor<AsRef<[u8]>>`][Cursor as Read]         | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`std::io::BufReader<Read>`]                             | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`std::net::TcpStream`][TcpStream as Read]               | ✅     |             | ✅          |         |          |             |
| [`&std::net::TcpStream`][&TcpStream as Read]             | ✅     |             | ✅          |         |          |             |
| [`std::os::unix::net::UnixStream`][UnixStream as Read]   | ✅     |             | ✅          |         |          |             |
| [`&std::os::unix::net::UnixStream`][&UnixStream as Read] | ✅     |             | ✅          |         |          |             |
| [`&std::sys::net::windows::Socket`]                      |       |             |            |         |          |             |
| [`std::io::PipeReader`]                                  | ✅     |             | ✅          |         | ✅        |             |
| [`&std::io::PipeReader`]                                 | ✅     |             | ✅          |         | ✅        |             |
| [`std::io::Stdin`]                                       | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`&std::io::Stdin`]                                      | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`std::io::StdinLock<'_>`]                               | ✅     | ✅           | ✅          | ✅       | ✅        | ✅           |
| [`std::io::stdio::StdinRaw`]                             | ✅     |             | ✅          |         | ✅        | ✅           |
| [`std::process::ChildStdout`]                            | ✅     |             | ✅          |         | ✅        |             |
| [`std::process::ChildStderr`]                            | ✅     |             | ✅          |         | ✅        |             |
| [`&std::sys::hermit::fd::FileDesc`]                      |       |             |            |         |          |             |
| [`std::sys::hermit::stdio::Stdin`]                       |       |             | ✅          |         |          |             |
| [`std::sys::sgx::stdio::Stdin`]                          |       |             |            |         |          |             |
| [`std::sys::solid::stdio::Stdin`]                        |       |             |            |         |          |             |
| [`std::sys::teeos::stdio::Stdin`]                        |       |             |            |         |          |             |
| [`std::sys::uefi::stdio::Stdin`]                         |       |             |            |         |          |             |
| [`std::sys::unix::fd::FileDesc`]                         | ✅     |             | ✅          |         |          |             |
| [`std::sys::unix::fs::CachedFileMetadata`][CFM as Read]  | ✅     |             | ✅          |         | ✅        | ✅           |
| [`std::sys::unix::stdio::Stdin`]                         | ✅     |             | ✅          |         |          |             |
| [`std::sys::unsupported::stdio::Stdin`]                  |       |             |            |         |          |             |
| [`std::sys::wasi::stdio::Stdin`]                         |       |             | ✅          |         |          |             |
| [`std::sys::windows::handle::Handle`]                    | ✅     |             | ✅          |         |          |             |
| [`std::sys::windows::stdio::Stdin`]                      |       |             |            |         |          |             |
| [`std::sys::xous::stdio::Stdin`]                         |       |             |            |         |          |             |
| [`std::sys::zkvm::stdio::Stdin`]                         |       |             |            |         |          |             |
| [`cargo::util::FileLock`][FileLock as Read]              |       |             |            |         |          |             |
| [`cargo::util::LimitErrorReader<Read>`]                  |       |             |            |         |          |             |

[`&mut Read`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L13
[`Box<Read>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L121
[&slice]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L236
[VecDeque as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L444
[`std::io::Chain<Read, Read>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L324
[`std::io::Take<Read>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/mod.rs#L2967
[Empty as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L60
[`std::io::Repeat`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L182
[File as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1235
[&File as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1096
[Arc<File> as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1281
[Cursor as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L324
[`std::io::BufReader<Read>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/buffered/bufreader.rs#L333
[TcpStream as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/net/tcp.rs#L623
[&TcpStream as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/net/tcp.rs#L662
[UnixStream as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/os/unix/net/stream.rs#L572
[&UnixStream as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/os/unix/net/stream.rs#L592
[`&std::sys::net::windows::Socket`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/net/connection/socket/windows.rs#L528
[`std::io::PipeReader`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/pipe.rs#L207
[`&std::io::PipeReader`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/pipe.rs#L187
[`std::io::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L437
[`&std::io::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L466
[`std::io::StdinLock<'_>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L503
[`std::io::stdio::StdinRaw`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L98
[`std::process::ChildStdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/process.rs#L407
[`std::process::ChildStderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/process.rs#L477
[`&std::sys::hermit::fd::FileDesc`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/hermit/fd.rs#L94
[`std::sys::hermit::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/hermit/stdio.rs#L18
[`std::sys::sgx::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/sgx/stdio.rs#L27
[`std::sys::solid::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/solid/stdio.rs#L15
[`std::sys::teeos::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/teeos/stdio.rs#L43
[`std::sys::uefi::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/uefi/stdio.rs#L49
[`std::sys::unix::fd::FileDesc`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/fd.rs#L578
[CFM as Read]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/fs.rs#L1972
[`std::sys::unix::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/stdio.rs#L16
[`std::sys::unsupported::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unsupported/stdio.rs#L13
[`std::sys::wasi::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/wasi/stdio.rs#L26
[`std::sys::windows::handle::Handle`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/windows/handle.rs#L336
[`std::sys::windows::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/windows/stdio.rs#L261
[`std::sys::xous::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/xous/stdio.rs#L16
[`std::sys::zkvm::stdio::Stdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/zkvm/stdio.rs#L15
[FileLock as Read]: https://github.com/rust-lang/cargo/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/src/cargo/util/flock.rs#L81
[`cargo::util::LimitErrorReader<Read>`]: https://github.com/rust-lang/cargo/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/src/cargo/util/io.rs#L16

## `Write`

```rust
pub trait Write {
    fn write(&mut self, buf: &[u8]) -> Result<usize>;
    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> Result<usize> { … }
    fn is_write_vectored(&self) -> bool { … }
    fn write_all(&mut self, buf: &[u8]) -> Result<()> { … }
    fn write_all_vectored(&mut self, bufs: &mut [IoSlice<'_>]) -> Result<()> { … }
    fn write_fmt(&mut self, fmt: fmt::Arguments<'_>) -> Result<()> { … }
    fn flush(&mut self) -> Result<()>;
}
```

Columns:
- `vectored`: `write_vectored` and `is_write_vectored`
- `all`: `write_all`
- `all_vectored`: `write_all_vectored`
- `fmt`: `write_fmt`

| `impl Write for T`                                        | `vectored` | `all` | `all_vectored` | `fmt` |
| --------------------------------------------------------- | ---------- | ----- | -------------- | ----- |
| [`&mut Write`]                                            | ✅          | ✅     |                | ✅     |
| [`Box<Write>`]                                            | ✅          | ✅     |                | ✅     |
| [`&mut [u8]`][&mut slice]                                 | ✅          | ✅     |                |       |
| [`Vec<u8>`]                                               | ✅          | ✅     |                |       |
| [`VecDeque<u8>`][VecDeque as Write]                       | ✅          | ✅     |                |       |
| [`std::io::Cursor<&mut [u8]>`][Cursor slice]              | ✅          |       |                |       |
| [`std::io::Cursor<&mut Vec<u8>>`]                         | ✅          |       |                |       |
| [`std::io::Cursor<Vec<u8>>`]                              | ✅          |       |                |       |
| [`std::io::Cursor<Box<[u8]>>`][Cursor box]                | ✅          |       |                |       |
| [`std::io::Cursor<[u8; N]>`][Cursor array]                | ✅          |       |                |       |
| [`core::io::BorrowedCursor<'_>`]                          |            |       |                |       |
| [`std::io::Empty`][Empty as Write]                        | ✅          |       |                |       |
| [`&std::io::Empty`]                                       | ✅          |       |                |       |
| [`std::io::Sink`]                                         | ✅          |       |                |       |
| [`&std::io::Sink`]                                        | ✅          |       |                |       |
| [`std::io::BufWriter<Write>`]                             | ✅          | ✅     |                |       |
| [`std::io::LineWriter<Write>`]                            | ✅          | ✅     | ✅              | ✅     |
| [`std::io::…::LineWriterShim<'_, Write>`]                 | ✅          | ✅     |                |       |
| [`std::fs::File`][File as Write]                          | ✅          |       |                |       |
| [`&std::fs::File`][&File as Write]                        | ✅          |       |                |       |
| [`Arc<std::fs::File>`][Arc<File> as Write]                | ✅          |       |                |       |
| [`std::net::TcpStream`][TcpStream as Write]               | ✅          |       |                |       |
| [`&std::net::TcpStream`][&TcpStream as Write]             | ✅          |       |                |       |
| [`std::os::unix::net::UnixStream`][UnixStream as Write]   | ✅          |       |                |       |
| [`&std::os::unix::net::UnixStream`][&UnixStream as Write] | ✅          |       |                |       |
| [`std::io::PipeWriter`]                                   | ✅          |       |                |       |
| [`&std::io::PipeWriter`]                                  | ✅          |       |                |       |
| [`std::io::Stdout`]                                       | ✅          | ✅     | ✅              | ✅     |
| [`&std::io::Stdout`]                                      | ✅          | ✅     | ✅              | ✅     |
| [`&std::io::StdoutLock<'_>`]                              | ✅          | ✅     | ✅              |       |
| [`std::io::stdio::StdoutRaw`]                             | ✅          | ✅     | ✅              | ✅     |
| [`std::io::Stderr`]                                       | ✅          | ✅     | ✅              | ✅     |
| [`&std::io::Stderr`]                                      | ✅          | ✅     | ✅              | ✅     |
| [`&std::io::StderrLock<'_>`]                              | ✅          | ✅     | ✅              |       |
| [`std::io::stdio::StderrRaw`]                             | ✅          | ✅     | ✅              | ✅     |
| [`std::process::ChildStdin`]                              | ✅          |       |                |       |
| [`&std::process::ChildStdin`]                             | ✅          |       |                |       |
| [`std::sys::hermit::stdio::Stdout`]                       | ✅          |       |                |       |
| [`std::sys::hermit::stdio::Stderr`]                       | ✅          |       |                |       |
| [`std::sys::sgx::stdio::Stdout`]                          |            |       |                |       |
| [`std::sys::sgx::stdio::Stderr`]                          |            |       |                |       |
| [`std::sys::sgx::abi::panic::SgxPanicOutput`]             |            |       |                |       |
| [`std::sys::solid::stdio::Stdout`]                        |            |       |                |       |
| [`std::sys::solid::stdio::Stderr`]                        |            |       |                |       |
| [`std::sys::solid::stdio::PanicOutput`]                   |            |       |                |       |
| [`std::sys::teeos::stdio::Stdout`]                        |            |       |                |       |
| [`std::sys::teeos::stdio::Stderr`]                        |            |       |                |       |
| [`std::sys::uefi::stdio::Stdout`]                         |            |       |                |       |
| [`std::sys::uefi::stdio::Stderr`]                         |            |       |                |       |
| [`std::sys::unix::fs::CachedFileMetadata`][CFM as Write]  | ✅          |       |                |       |
| [`std::sys::unix::stdio::Stdout`]                         | ✅          |       |                |       |
| [`std::sys::unix::stdio::Stderr`]                         | ✅          |       |                |       |
| [`std::sys::unsupported::stdio::Stdout`]                  |            |       |                |       |
| [`std::sys::unsupported::stdio::Stderr`]                  |            |       |                |       |
| [`std::sys::wasi::stdio::Stdout`]                         | ✅          |       |                |       |
| [`std::sys::wasi::stdio::Stderr`]                         | ✅          |       |                |       |
| [`std::sys::windows::stdio::Stdout`]                      |            |       |                |       |
| [`std::sys::windows::stdio::Stderr`]                      |            |       |                |       |
| [`std::sys::xous::stdio::Stdout`]                         |            |       |                |       |
| [`std::sys::xous::stdio::Stderr`]                         |            |       |                |       |
| [`std::sys::xous::stdio::PanicWriter`]                    |            |       |                |       |
| [`std::sys::xkvm::stdio::Stdout`]                         |            |       |                |       |
| [`std::sys::xkvm::stdio::Stderr`]                         |            |       |                |       |
| [`test::console::OutputLocation<Write>`]                  |            |       |                |       |
| [`test::term::WinConsole<Write>`]                         |            |       |                |       |
| [`test::term::TerminfoTerminal<Write>`]                   |            |       |                |       |
| [`cargo::util::FileLock`][FileLock as Write]              |            |       |                |       |
| [`proc_macro::bridge::buffer::Buffer`]                    |            | ✅     |                |       |
| [`rust-installer::compression::CombinedEncoder`]          |            | ✅     |                |       |
| [`rustc_errors::emitter::Buffy`]                          |            |       |                |       |
| [`rustc_errors::json::…::BufWriter`]                      |            |       |                |       |

[`&mut Write`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L54
[`Box<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L162
[&mut slice]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L366
[`Vec<u8>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L408
[VecDeque as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L551
[Cursor slice]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L566
[`std::io::Cursor<&mut Vec<u8>>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L589
[`std::io::Cursor<Vec<u8>>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L613
[Cursor box]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L637
[Cursor array]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/cursor.rs#L663
[`core::io::BorrowedCursor<'_>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/impls.rs#L586
[Empty as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L104
[`&std::io::Empty`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L128
[`std::io::Sink`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L285
[`&std::io::Sink`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/util.rs#L309
[`std::io::BufWriter<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/buffered/bufwriter.rs#L525
[`std::io::LineWriter<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/buffered/linewriter.rs#L191
[`std::io::…::LineWriterShim<'_, Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/buffered/linewritershim.rs#L54
[File as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1257
[&File as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1164
[Arc<File> as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/fs.rs#L1303
[TcpStream as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/net/tcp.rs#L642
[&TcpStream as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/net/tcp.rs#L681
[UnixStream as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/os/unix/net/stream.rs#L612
[&UnixStream as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/os/unix/net/stream.rs#L632
[`std::io::PipeWriter`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/pipe.rs#L245
[`&std::io::PipeWriter`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/pipe.rs#L227
[`std::io::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L745
[`&std::io::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L771
[`&std::io::StdoutLock<'_>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L803
[`std::io::stdio::StdoutRaw`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L125
[`std::io::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L977
[`&std::io::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L1003
[`&std::io::StderrLock<'_>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L1035
[`std::io::stdio::StderrRaw`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/io/stdio.rs#L157
[`std::process::ChildStdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/process.rs#L321
[`&std::process::ChildStdin`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/process.rs#L341
[`std::sys::hermit::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/hermit/stdio.rs#L41
[`std::sys::hermit::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/hermit/stdio.rs#L68
[`std::sys::sgx::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/sgx/stdio.rs#L39
[`std::sys::sgx::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/sgx/stdio.rs#L55
[`std::sys::sgx::abi::panic::SgxPanicOutput`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/sgx/abi/panic.rs#L29
[`std::sys::solid::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/solid/stdio.rs#L27
[`std::sys::solid::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/solid/stdio.rs#L44
[`std::sys::solid::stdio::PanicOutput`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/solid/stdio.rs#L61
[`std::sys::teeos::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/teeos/stdio.rs#L55
[`std::sys::teeos::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/teeos/stdio.rs#L71
[`std::sys::uefi::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/uefi/stdio.rs#L110
[`std::sys::uefi::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/uefi/stdio.rs#L129
[CFM as Write]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/fs.rs#L1993
[`std::sys::unix::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/stdio.rs#L41
[`std::sys::unix::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unix/stdio.rs#L69
[`std::sys::unsupported::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unsupported/stdio.rs#L25
[`std::sys::unsupported::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/unsupported/stdio.rs#L41
[`std::sys::wasi::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/wasi/stdio.rs#L54
[`std::sys::wasi::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/wasi/stdio.rs#L85
[`std::sys::windows::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/windows/stdio.rs#L439
[`std::sys::windows::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/windows/stdio.rs#L455
[`std::sys::xous::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/xous/stdio.rs#L28
[`std::sys::xous::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/xous/stdio.rs#L55
[`std::sys::xous::stdio::PanicWriter`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/xous/stdio.rs#L88
[`std::sys::xkvm::stdio::Stdout`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/zkvm/stdio.rs#L27
[`std::sys::xkvm::stdio::Stderr`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/std/src/sys/pal/zkvm/stdio.rs#L45
[`test::console::OutputLocation<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/test/src/console.rs#L28
[`test::term::WinConsole<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/test/src/term/win.rs#L141
[`test::term::TerminfoTerminal<Write>`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/test/src/term/terminfo/mod.rs#L193
[FileLock as Write]: https://github.com/rust-lang/cargo/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/src/cargo/util/flock.rs#L93
[`proc_macro::bridge::buffer::Buffer`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/library/proc_macro/src/bridge/buffer.rs#L102
[`rust-installer::compression::CombinedEncoder`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/src/tools/rust-installer/src/compression.rs#L213
[`rustc_errors::emitter::Buffy`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/compiler/rustc_errors/src/emitter.rs#L3325
[`rustc_errors::json::…::BufWriter`]: https://github.com/rust-lang/rust/blob/43ca9d18e333797f0aa3b525501a7cec8d61a96b/compiler/rustc_errors/src/json.rs#L322

**Comment by thaliaarchi** [comment]
FYI, https://github.com/rust-lang/rust/issues/140771 adds the following methods, which should be incorporated into the table:

```rust
// std::os::unix::fs
trait FileExt {
    fn read_buf_at(&self, buf: BorrowedCursor<'_>, offset: u64) -> io::Result<()>;
    fn read_buf_exact_at(&self, buf: BorrowedCursor<'_>, offset: u64) -> io::Result<()>;
}

// std::os::windows::fs
trait FileExt {
    fn seek_read_buf(&self, buf: BorrowedCursor<'_>, offset: u64) -> io::Result<()>;
}
```
